File: frontend/src/types/migration.ts
export type MigrationStatus =
  | "pending"
  | "analyzing"
  | "ready"
  | "applying"
  | "completed"
  | "failed";

export interface MigrationListItem {
  id: string;
  source_device_id: string;
  target_device_id: string;
  source_vendor: string;
  target_vendor: string;
  status: MigrationStatus;
  ai_level: number;
  created_at: string;
  updated_at: string;
}

export interface ParsedVlan {
  name: string | null;
}

export interface ParsedInterface {
  name: string;
  mode: "trunk" | "access" | "hybrid";
  pvid: string | null;
  tagged_vlans: string[];
  description: string | null;
  port_type: "ethernet" | "fiber" | "lag" | "vlan" | "unknown";
}

export interface L3Interface {
  vlan_id: string;
  ip: string;
  mask: string;
}

export interface MigrationPlan {
  hostname: string | null;
  stp_mode: string | null;
  vlans: Record<string, ParsedVlan>;
  interfaces: ParsedInterface[];
  l3_interfaces: L3Interface[];
  warnings: string[];
}

export interface Migration {
  id: string;
  tenant_id: string;
  source_device_id: string;
  target_device_id: string;
  source_vendor: string;
  target_vendor: string;
  status: MigrationStatus;
  source_config_raw: string | null;
  target_config_raw: string | null;
  migration_plan: MigrationPlan | null;
  port_mapping: Record<string, string> | null;
  commands_preview: string | null;
  warnings: string[] | null;
  error_message: string | null;
  ai_level: number;
  created_at: string;
  updated_at: string;
}

export interface MigrationCreate {
  source_device_id: string;
  target_device_id: string;
  ai_level?: number;
}

export interface PortMappingUpdate {
  port_mapping: Record<string, string>;
}

export interface RegenerateRequest {
  port_mapping?: Record<string, string>;
}

export interface InterfaceAdd {
  name: string;
  target_name: string;
  mode?: string;
  pvid?: string | null;
  tagged_vlans?: string[];
  description?: string | null;
  port_type?: string;
}

File: backend/app/schemas/migration.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class MigrationCreate(BaseModel):
    source_device_id: str
    target_device_id: str
    ai_level: int = 2


class PortMappingUpdate(BaseModel):
    port_mapping: dict[str, str]  # { "source_port_name": "target_port_name" }


class CommandsUpdate(BaseModel):
    commands_preview: str


class RegenerateRequest(BaseModel):
    port_mapping: dict[str, str] | None = None


class InterfaceAdd(BaseModel):
    name: str
    target_name: str
    mode: str = "access"
    pvid: str | None = None
    tagged_vlans: list[str] = []
    description: str | None = None
    port_type: str = "ethernet"


class MigrationRead(BaseModel):
    id: UUID
    tenant_id: UUID
    source_device_id: UUID
    target_device_id: UUID
    source_vendor: str
    target_vendor: str
    status: str
    source_config_raw: str | None
    target_config_raw: str | None
    migration_plan: dict | None
    port_mapping: dict | None
    commands_preview: str | None
    warnings: list | None
    error_message: str | None
    ai_level: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MigrationListItem(BaseModel):
    id: UUID
    source_device_id: UUID
    target_device_id: UUID
    source_vendor: str
    target_vendor: str
    status: str
    ai_level: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
